androidApp/src/main/java/com/example/tasky/android/agenda/viewmodel/AgendaDetailsViewModel.kt

```diff
@@ -112,25 +117,67 @@ class AgendaDetailsViewModel(
         minute: Int,
     ) {
         val agendaItem = screenStateFlow.value.agendaItem
-        val timeMilli = hour.hours.toLong(DurationUnit.MILLISECONDS) + minute.minutes.toLong(DurationUnit.MILLISECONDS)
-        val oneDayMilli = 1.days.toLong(DurationUnit.MILLISECONDS)
-        val newDateTime = agendaItem.getStartTime() / oneDayMilli * oneDayMilli + timeMilli
+        val newDateTime =
+            LocalDateTime(agendaItem.getStartTime().toLocalDateTime().date, LocalTime(hour = hour, minute = minute))
+                .toInstant(
+                    TimeZone.currentSystemDefault(),
+                ).toEpochMilliseconds()
         val remindAtDiff = agendaItem.getStartTime() - agendaItem.remindAt
         _screenStateFlow.update {
             it.copy(agendaItem = it.agendaItem.copy(startTime = newDateTime, remindAt = newDateTime - remindAtDiff))
         }
     }
 
-    private fun updateStartDate(newDate: Long) {
+    private fun updateStartDate(newDate: LocalDateTime) {
         val agendaItem = screenStateFlow.value.agendaItem
-        val oneDayMilli = 1.days.toLong(DurationUnit.MILLISECONDS)
-        val newDateTime = newDate + agendaItem.getStartTime() % oneDayMilli
+        val newDateTime =
+            LocalDateTime(newDate.date, agendaItem.getStartTime().toLocalDateTime().time)
+                .toInstant(
+                    TimeZone.currentSystemDefault(),
+                ).toEpochMilliseconds()
         val remindAtDiff = agendaItem.getStartTime() - agendaItem.remindAt
         _screenStateFlow.update {
             it.copy(agendaItem = it.agendaItem.copy(startTime = newDateTime, remindAt = newDateTime - remindAtDiff))
         }
     }
 
+    private fun updateEndTime(
+        hour: Int,
+        minute: Int,
+    ) {
+        val agendaItem = screenStateFlow.value.agendaItem
+
+        if (agendaItem !is Event) {
+            return
+        }
+
+        val newDateTime =
+            LocalDateTime(agendaItem.to.toLocalDateTime().date, LocalTime(hour = hour, minute = minute))
+                .toInstant(
+                    TimeZone.currentSystemDefault(),
+                ).toEpochMilliseconds()
+        _screenStateFlow.update {
+            it.copy(agendaItem = agendaItem.copy(to = newDateTime))
+        }
+    }
+
+    private fun updateEndDate(newDate: LocalDateTime) {
+        val agendaItem = screenStateFlow.value.agendaItem
+
+        if (agendaItem !is Event) {
+            return
+        }
```

Comment:
I personally added a sealed interface like the following one which contains the UI state that is specific to a certain item type:
```
sealed interface AgendaItemDetails {
    data class Event(
        val toTime: LocalDateTime,
        val attendees: List<Attendee>,
        val photos: List<Photo>,
        // ...
    ): AgendaItemDetails
    
    data class Task(
        val isDone: Boolean
    ): AgendaItemDetails
    
    data object Reminder: AgendaItemDetails
}
```
That way, you communicate by design which fields are specific to which item type (and you don't need to add those safety checks) and you just need type-checks when updating properties that are individual to one of the agenda item types, but you don't need this for updating shared properties anymore like the title:
```
data class AgendaDetailState(
    val title: String = "",
    val description: String = "",
    val time: LocalDateTime = LocalDateTime.now(),
    val details: AgendaItemDetails? = null
)
```
Updating then works like this:
```
// Updating a shared property
state = state.copy(title = "new title")

// Updating an individual property
state = state.copy(
    details = detailsAsEvent()?.copy(attendees = newAttendees)
)
```
And this function is quite handy:
```
private fun updateDetailsIfEvent(update: (AgendaItemDetails.Event) -> AgendaItemDetails.Event): AgendaItemDetails {
    return when (val details = state.value.details) {
        is AgendaItemDetails.Event -> update(details)
        else -> details
    }
}
```
That way you could update an event's unique state like this with the following approach:
```
_state.update {
    it.copy(
        details = updateDetailsIfEvent {
            it.copy(photos = newPhotos)
        }
    )
}
```